header-only/layout-builder/awkward/LayoutBuilder.h

```diff
@@ -1211,17 +1219,23 @@ namespace awkward {
       /// @brief Checks for validity and consistency.
       bool
       is_valid(std::string& error) const noexcept {
-        if (content_.length() != index_.length()) {
-          std::stringstream out;
-          out << "Indexed node" << id_ << " has content length "
-              << content_.length() << " but index has length " << index_.length()
-              << "\n";
-          error.append(out.str());
 
-          return false;
-        } else {
-          return content_.is_valid(error);
+        std::unique_ptr<PRIMITIVE[]> ptr(new PRIMITIVE[index_.length()]);
+        index_.concatenate(ptr.get());
+        // check that each element of index_ is >= 0 and < content_.length()
```

Comment:
This would also be possible without the $\mathcal{O}(n)$ check by keeping track of the maximum index value as you go along, instead of checking all the values at the end.

Reply:
is there an alternative to `concatenate` that I can use to achieve this?

Reply:
or do you mean modify `append_index` to keep track of max index as a private member of the class?

Reply:
I wasn't able to select the code that I wanted on my phone, so I commented on the first line. I meant this whole block:

```c++
        for (size_t i = 0; i < index_.length(); i++) {
          if (ptr.get()[i] < 0 ||
              ptr.get()[i] >= content_.length()) {
            std::stringstream out;
            out << "Indexed node" << id_ << " has index " << ptr.get()[i]
                << " at position " << i << " but content has length "
                << content_.length() << "\n";
            error.append(out.str());

            return false;
          }
```

At the time when `is_valid` is called, this runs through the whole `index_` to see if any of its values are `>= content_.length()`. That's a $\mathcal{O}(n)$ test (albeit a compiled one, so not terrible). Instead of that extra run through memory addresses that have probably been evicted from CPU cache and would have to be loaded again, this class could have kept a

```c++
  int64_t max_index_;
```

member that is always equal to the largest `index_` that has been seen so far. So, everywhere that `index_` is updated (only a few member functions do that), check to see if the new `index_` is greater than `max_index_`, and if it is, update `max_index_`. Then when you come to the `is_valid` function, you only need to verify that

```c++
  max_index_ < content_.length()
```

which is a $\mathcal{O}(1)$ check. We've traded a little memory (8 bytes) for some CPU time (running through all $n$ values in `index_`).

Both IndexedArray and IndexedOptionArray can benefit from this, both taking on a single new 8 byte member, and UnionArray can as well, though it needs to take on an 8 byte member for every one of its `contents_` vectors.

If `is_valid` is run frequently, like once per `append`, then this should definitely be fixed. (Once per `append`, which is $\mathcal{O}(n)$, would make the whole process $\mathcal{O}(n^2)$.) But I think `is_valid` is run only once at the end of array-building, in which case it doesn't matter a lot. It's one unnecessary pass over the data, but Awkward Arrays in Python do a lot of passes over the data.

Reply:
`is_valid` is called just once after loading the data:

https://gist.github.com/zonca/8045dc6294e2b9ff4e053a339a3a77fb#file-simple_enum-cpp-L115-L116

however, is `concatenate` making an entire copy of the data? If so, it is probably better to implement `max_index_` to avoid memory overhead.

Reply:
Yes, `concatenate` does copy the entire data. The data are filled in fixed-size panels, which allows them to grow indefinitely, but Awkward Array needs buffers to be (a) contiguous and (b) owned by Python, so that's why we have a concatenate-while-copying step.

Earlier, we had the data grow as `std::vectors`, which reallocates logarithmically many times (because it allocates a bigger buffer each time), but since we had to copy the data to make it owned by NumPy arrays _anyway_, we gave up on trying to keep them contiguous while filling. That's how we ended up with our current pattern.

Reply:
ok, implemented